starter-ui/src/main/java/fish/payara/starter/resources/ApplicationGenerator.java

```diff
@@ -233,40 +239,38 @@ private void invokeMavenArchetype(String archetypeGroupId, String archetypeArtif
         }
     }
 
-    private File zipDirectory(File directory, File destinaton) throws IOException {
-        File zipFile = new File(destinaton, directory.getName() + ZIP_EXTENSION);
-        try (ZipOutputStream zos = new ZipOutputStream(new FileOutputStream(zipFile))) {
-            zipFile(directory, directory.getName(), zos);
-        }
-        return zipFile;
-    }
+    private File zipDirectory(File directory, File destinatonDir) throws IOException {
+        Set<String> executables = Set.of("/mvnw", "/gradlew");
+        Path directoryPath = directory.toPath();
+        File zipFile = new File(destinatonDir, directory.getName() + ZIP_EXTENSION);
+        Map<String, String> zipProperties = Map.of("create", "true");
 
-    private void zipFile(File fileToZip, String fileName, ZipOutputStream zipOut) throws IOException {
-        if (fileToZip.isDirectory()) {
-            if (fileName.endsWith("/")) {
-                zipOut.putNextEntry(new ZipEntry(fileName));
-                zipOut.closeEntry();
-            } else {
-                zipOut.putNextEntry(new ZipEntry(fileName + "/"));
-                zipOut.closeEntry();
-            }
-
-            File[] children = fileToZip.listFiles();
-            for (File childFile : children) {
-                zipFile(childFile, fileName + "/" + childFile.getName(), zipOut);
-            }
-        } else {
-            try (FileInputStream fis = new FileInputStream(fileToZip)) {
-                ZipEntry zipEntry = new ZipEntry(fileName);
-                zipOut.putNextEntry(zipEntry);
+        // Create a ZIP file
+        try (FileSystem zipFs = FileSystems.newFileSystem(URI.create("jar:" + zipFile.toPath().toUri()), zipProperties)) {
+            // Walk through the directory tree and copy each file into the ZIP filesystem
+            Files.walkFileTree(directoryPath, new SimpleFileVisitor<>() {
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                    Path destFile = zipFs.getPath("/").resolve(directoryPath.relativize(file).toString());
+                    Files.createDirectories(destFile.getParent());
+                    Files.copy(file, destFile, StandardCopyOption.REPLACE_EXISTING);
+                    if (executables.contains(destFile.toString())) {
```

Comment:
Another check for Windows must be added `if (!System.getProperty("os.name").toLowerCase().contains("win"))` as it throws UnsupportedOperationException in Windows.

````
java.lang.UnsupportedOperationException
        at java.base/sun.nio.fs.WindowsFileSystemProvider.readAttributes(WindowsFileSystemProvider.java:198)
        at java.base/java.nio.file.Files.readAttributes(Files.java:1851)
        at java.base/java.nio.file.Files.getPosixFilePermissions(Files.java:2125)
        at fish.payara.starter.resources.ApplicationGenerator$1.visitFile(ApplicationGenerator.java:263)
        at fish.payara.starter.resources.ApplicationGenerator$1.visitFile(ApplicationGenerator.java:254)
        ````